libs/hmpb/src/ballot_rotation.ts

```diff
@@ -0,0 +1,191 @@
+import { BallotTemplateId } from './ballot_templates';
+
+/**
+ * Helper function that orders candidates alphabetically by name,
+ * rotating the alphabet so that it starts with the first letter of the precinct name.
+ * For example, if the precinct is "Central", candidates starting with C come first,
+ * then D, E, ..., Z, A, B.
+ *
+ * This rotation logic is only intended for demonstration purposes and testing.
+ */
+export function getCandidateOrderingByPrecinctAlphabetical({
+  contests,
+  precincts,
+  precinctsOrSplitIds,
+  districtIds,
+}: RotationParams): CandidateOrderingSet[] {
+  const ballotStyleContests = contests.filter((contest) =>
+    districtIds.includes(contest.districtId)
+  );
+
+  // Generate a separate ordering for each precinct/split
+  return precinctsOrSplitIds.map((precinctOrSplit) => {
+    const precinct = find(
+      precincts,
+      (p) => p.id === precinctOrSplit.precinctId
+    );
+    const precinctName = precinct.name;
+    const firstLetter = precinctName.charAt(0).toUpperCase();
+
+    const orderedContests: Record<ContestId, DisplayCandidate[]> = {};
+
+    for (const contest of ballotStyleContests) {
+      switch (contest.type) {
+        case 'yesno':
+          // do nothing
+          break;
+        case 'candidate': {
+          // Sort candidates by name, rotated by precinct's first letter
+          let sortedCandidates = [...contest.candidates].sort((a, b) =>
+            a.name.localeCompare(b.name)
+          );
+
+          // rotate so candidates starting with letters at or after the precinct letter come first
+          const startIndex = sortedCandidates.findIndex(
+            (c) => (c.name.charAt(0) || '').toUpperCase() >= firstLetter
+          );
+
+          if (startIndex > 0) {
+            sortedCandidates = [
+              ...sortedCandidates.slice(startIndex),
+              ...sortedCandidates.slice(0, startIndex),
+            ];
+          }
+
+          orderedContests[contest.id] = sortedCandidates.map((candidate) =>
+            typedAs<DisplayCandidate>({
+              id: candidate.id,
+            })
+          );
+          break;
+        }
+        default: {
+          /* istanbul ignore next - @preserve */
+          return throwIllegalValue(contest, 'type');
+        }
+      }
+    }
+
+    return {
+      precinctsOrSplits: [precinctOrSplit],
+      orderedContests,
+    };
+  });
+}
+
+/**
+ * Generates ordered contests for VxDefault ballot template.
+ * This template does not apply any rotation.
+ * Returns a single set with the contests in their original order.
+ */
+function getDefaultCandidateOrdering({
+  contests,
+  precinctsOrSplitIds,
+  districtIds,
+}: RotationParams): CandidateOrderingSet[] {
+  // Get contests for this ballot style (filtered by district)
+  const ballotStyleContests = contests.filter((contest) =>
+    districtIds.includes(contest.districtId)
+  );
```

Comment:
I kind of feel like this should be done higher up in the call stack, since it applies to all ballot rotation. I.e. maybe when generating the ballot style, we should filter which contests apply and then pass that into the rotation logic. We could also ensure that only candidate contests get passed in, which would simplify some of the rotation logic.